### PR TITLE
Smaclachlan/solver stats

### PR DESCRIPTION
**Pull request by rckirby**
We provide some simple solver statistics -- each kind of time stepper tracks the number of steps it has taken and the total number of linear/nonlinear iterations it took over those time steps.  Different methods report a different kind of stats (e.g. IMEX needs to report iterations separately for iterator/propagator).

### CODE REVIEW

File: demos/heat/demo_heat_dirk.py.rst [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  from irksome import Alexander, Dt, TimeStepper
+
+And we will need a little bit of UFL to support using the method of
+manufactured solutions::
+
+  from ufl.algorithms.ad import expand_derivatives
+
+We will create the Butcher tableau for the lowest-order Gauss-Legendre
+Runge-Kutta method, which is more commonly known as the implicit
+midpoint rule::
```
Comment (ScottMacLachlan) [inline demos/heat/demo_heat_dirk.py.rst:44]: This is not true in this demo

File: demos/heat/demo_heat_dirk.py.rst [hunk 2]
```diff
@@ -0,0 +1,126 @@
+Solving the Heat Equation with Irksome
+======================================
+
```
Comment (ScottMacLachlan) [inline demos/heat/demo_heat_dirk.py.rst:3]: Should we note that this is the DIRK demo in the title block?  Seems clear from the filename, of course, but would be nice to emphasize the distinction a bit more.

File: demos/monodomain/demo_monodomain_FHN.py.rst [hunk 3]
```diff
@@ -0,0 +1,121 @@
+Solving monodomain equations with Fitzhugh-Nagumo reaction and an IMEX method
```
Comment (ScottMacLachlan) [inline demos/monodomain/demo_monodomain_FHN.py.rst:1]: As a demo, this leaves a bit to be desired in the documentation department

File: demos/monodomain/demo_monodomain_FHN.py.rst [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  butcher_tableau = RadauIIA(2)
+  ns = butcher_tableau.num_stages
+
+  uu = Function(Z)
+  vu, vc = TestFunctions(Z)
+  uu.sub(0).interpolate(InitialPotential)
+  uu.sub(1).interpolate(InitialCell)
+
+  (uCurr, cCurr) = split(uu)
+
```
Comment (ScottMacLachlan) [inline demos/monodomain/demo_monodomain_FHN.py.rst:50]: Definitely need to document/explain the IMEX splitting here

**Review by ScottMacLachlan — CHANGES_REQUESTED** [review]
Aside from some small details in the demos, this LGTM.

On the one hand, we still don't have a demo for stage type `value`.  On the other hand, that's a change in a single kwarg from the stage type `deriv` demos, so maybe we don't need one?

File: demos/heat/demo_heat_dirk.py.rst (hunk 1 above)
Comment (rckirby) [reply demos/heat/demo_heat_dirk.py.rst:44]: You're right.

File: demos/heat/demo_heat_dirk.py.rst (hunk 2 above)
Comment (rckirby) [reply demos/heat/demo_heat_dirk.py.rst:3]: Ok

File: demos/monodomain/demo_monodomain_FHN.py.rst [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+   c_t = \epsilon( u + \beta - \gamma c)
+
+for certain positive parameters :math:`\beta` and :math:`\gamma`, and the current takes the form of:
+
+.. math::
+
+   I_{ion}(u, c) = \tfrac{1}{\epsilon} \left( u - \tfrac{u^3}{3} - c \right)
+
+so that we have an overall system of two equations.  One of them is linear but stiff/diffusive, and theo ther is nonstiff but nonlinear.  This combination makes the system a good candidate for additive partitioning/IMEX-type methods.
```
Comment (ScottMacLachlan) [inline demos/monodomain/demo_monodomain_FHN.py.rst:23]: Typo: `theo ther` -> `the other`

File: demos/monodomain/demo_monodomain_FHN.py.rst (hunk 4 above)
Comment (ScottMacLachlan) [reply demos/monodomain/demo_monodomain_FHN.py.rst:72]: Yep - aside from the typo, this looks good.  Thanks!

**Review by ScottMacLachlan — APPROVED** [review]
Small typo, otherwise looks good!

**Pull request merged**